Regression/ICE: moving both a `Copy` type and a `!Copy` type into `async move || { ... }` (`async_closure`)

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

I tried this code (minimized):

```rust
#![feature(async_closure)]

fn foo() {
    // no ICE if removed or made !Copy
    let bar = 0u8;
    // no ICE if removed or made Copy
    let baz = "".to_owned();
    async move || {
        let _ = (bar, baz);
    };
}

fn main() {}
```

I expected to see this happen: Compiles

Instead, this happened **(`nightly-2024-02-09`)**:
```rust
error[E0597]: `bar` does not live long enough
 --> src/main.rs:7:18
  |
6 |       async move || {
  |  ___________________-
7 | |         let _ = (bar, baz);
  | |                  ^^^ borrowed value does not live long enough
8 | |     };
  | |     -- borrow later used here
  | |     |
  | |_____`bar` dropped here while still borrowed
  |       value captured here by coroutine
```

Instead, this happened **(latest nightly)**: ICE
```rust
   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at compiler/rustc_mir_transform/src/coroutine/by_move_body.rs:140:21:
assertion `left != right` failed: `FnOnce` coroutine-closures return coroutines that capture from their body; it will always result in a borrowck error!
  left: FnOnce
 right: FnOnce
```

### Versions

#### working -> error
```console
searched nightlies: from nightly-2023-04-25 to nightly-2024-04-20
regressed nightly: nightly-2024-02-07
searched commit range: https://github.com/rust-lang/rust/compare/f067fd6084d750f3797f54b71771c5dbc149726f...256b6fb19a2c018eaad4806d2369d1f6a71fc6ec
regressed commit: https://github.com/rust-lang/rust/commit/4a2fe4491ea616983a0cf0cbbd145a39768f4e7a

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.6</summary>
```

Related: https://github.com/rust-lang/rust/pull/120361
Mentioning @compiler-errors :wink:

#### error -> ICE
```console
********************************************************************************
Regression in 1921968cc5403892739b43bdefe793a130badd15
********************************************************************************
```

Best guess: https://github.com/rust-lang/rust/pull/123350

<details><summary>Backtrace</summary>
<p>

```
 Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at compiler/rustc_mir_transform/src/coroutine/by_move_body.rs:140:21:
assertion `left != right` failed: `FnOnce` coroutine-closures return coroutines that capture from their body; it will always result in a borrowck error!
  left: FnOnce
 right: FnOnce
stack backtrace:
   0:     0x7f5b0babe015 - std::backtrace_rs::backtrace::libunwind::trace::h202d45d8fd20a824
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x7f5b0babe015 - std::backtrace_rs::backtrace::trace_unsynchronized::h1daee13ebb59eeb4
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f5b0babe015 - std::sys_common::backtrace::_print_fmt::hf1c1d6c712400d1b
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f5b0babe015 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha6cb3dac0d776969
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f5b0bb0d27b - core::fmt::rt::Argument::fmt::hb86c5e5f8ba90810
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/core/src/fmt/rt.rs:165:63
   5:     0x7f5b0bb0d27b - core::fmt::write::h5b057186bbc5fec2
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/core/src/fmt/mod.rs:1157:21
   6:     0x7f5b0bab2bbf - std::io::Write::write_fmt::h62aa7b9115cec986
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/io/mod.rs:1832:15
   7:     0x7f5b0babddee - std::sys_common::backtrace::_print::hdfcedaa64cb5957c
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f5b0babddee - std::sys_common::backtrace::print::h128dd53d70614027
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f5b0bac0759 - std::panicking::default_hook::{{closure}}::h25cd5b2272122564
  10:     0x7f5b0bac049d - std::panicking::default_hook::h3f62a662b49f346a
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/panicking.rs:298:9
  11:     0x7f5b0eb62d0c - std[f6dbfb092428beee]::panicking::update_hook::<alloc[1dedc0b6462c0e58]::boxed::Box<rustc_driver_impl[a06908b56a08a5f3]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f5b0bac0e56 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h195b48b24ae880e2
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/alloc/src/boxed.rs:2036:9
  13:     0x7f5b0bac0e56 - std::panicking::rust_panic_with_hook::h906fa2f149cd991b
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/panicking.rs:799:13
  14:     0x7f5b0bac0c04 - std::panicking::begin_panic_handler::{{closure}}::hf5f40df14bc3ecc3
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/panicking.rs:664:13
  15:     0x7f5b0babe4d9 - std::sys_common::backtrace::__rust_end_short_backtrace::h6caf7624fbbe1a8f
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f5b0bac0937 - rust_begin_unwind
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/panicking.rs:652:5
  17:     0x7f5b0bb09743 - core::panicking::panic_fmt::h24ccdca200c3cead
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/core/src/panicking.rs:72:14
  18:     0x7f5b0bb09d0f - core::panicking::assert_failed_inner::h154a63e2b6258de7
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/core/src/panicking.rs:404:23
  19:     0x7f5b0f161938 - core[3cc7221ca1340ad0]::panicking::assert_failed::<rustc_type_ir[f35b0a68110f4e13]::ClosureKind, rustc_type_ir[f35b0a68110f4e13]::ClosureKind>
  20:     0x7f5b0e383740 - <rustc_mir_transform[2378a63253b3ac0c]::coroutine::by_move_body::ByMoveBody as rustc_middle[fb5dda08dce04472]::mir::MirPass>::run_pass
  21:     0x7f5b0ff3394d - rustc_mir_transform[2378a63253b3ac0c]::pass_manager::run_passes_inner
  22:     0x7f5b102521a8 - rustc_query_impl[ffc1dc643be498fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ffc1dc643be498fd]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 8usize]>>
  23:     0x7f5b10348771 - rustc_query_system[6eb3337d20c14267]::query::plumbing::try_execute_query::<rustc_query_impl[ffc1dc643be498fd]::DynamicConfig<rustc_query_system[6eb3337d20c14267]::query::caches::VecCache<rustc_span[413b5bd4996aee21]::def_id::LocalDefId, rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[ffc1dc643be498fd]::plumbing::QueryCtxt, false>
  24:     0x7f5b1034828c - rustc_query_impl[ffc1dc643be498fd]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7f5b103340b0 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor>::visit_inner_body
  26:     0x7f5b0d506cd4 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor as rustc_middle[fb5dda08dce04472]::thir::visit::Visitor>::visit_expr
  27:     0x7f5b103341c1 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor>::visit_inner_body
  28:     0x7f5b0d506cd4 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor as rustc_middle[fb5dda08dce04472]::thir::visit::Visitor>::visit_expr
  29:     0x7f5b1077d119 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor as rustc_middle[fb5dda08dce04472]::thir::visit::Visitor>::visit_block
  30:     0x7f5b0d507fd0 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor as rustc_middle[fb5dda08dce04472]::thir::visit::Visitor>::visit_expr
  31:     0x7f5b0d506cd4 - <rustc_mir_build[a31e700f633e9780]::check_unsafety::UnsafetyVisitor as rustc_middle[fb5dda08dce04472]::thir::visit::Visitor>::visit_expr
  32:     0x7f5b10331324 - rustc_mir_build[a31e700f633e9780]::check_unsafety::check_unsafety
  33:     0x7f5b10330dd1 - rustc_query_impl[ffc1dc643be498fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ffc1dc643be498fd]::query_impl::check_unsafety::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 0usize]>>
  34:     0x7f5b1032fe4d - rustc_query_system[6eb3337d20c14267]::query::plumbing::try_execute_query::<rustc_query_impl[ffc1dc643be498fd]::DynamicConfig<rustc_query_system[6eb3337d20c14267]::query::caches::VecCache<rustc_span[413b5bd4996aee21]::def_id::LocalDefId, rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[ffc1dc643be498fd]::plumbing::QueryCtxt, false>
  35:     0x7f5b1032faff - rustc_query_impl[ffc1dc643be498fd]::query_impl::check_unsafety::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7f5b103282e5 - rustc_interface[7ce95aec19cf85e]::passes::analysis
  37:     0x7f5b103273d5 - rustc_query_impl[ffc1dc643be498fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ffc1dc643be498fd]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 1usize]>>
  38:     0x7f5b10d2a9e5 - rustc_query_system[6eb3337d20c14267]::query::plumbing::try_execute_query::<rustc_query_impl[ffc1dc643be498fd]::DynamicConfig<rustc_query_system[6eb3337d20c14267]::query::caches::SingleCache<rustc_middle[fb5dda08dce04472]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[ffc1dc643be498fd]::plumbing::QueryCtxt, false>
  39:     0x7f5b10d2a749 - rustc_query_impl[ffc1dc643be498fd]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f5b10b8a98e - rustc_interface[7ce95aec19cf85e]::interface::run_compiler::<core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>, rustc_driver_impl[a06908b56a08a5f3]::run_compiler::{closure#0}>::{closure#1}
  41:     0x7f5b10b77327 - std[f6dbfb092428beee]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[7ce95aec19cf85e]::util::run_in_thread_with_globals<rustc_interface[7ce95aec19cf85e]::util::run_in_thread_pool_with_globals<rustc_interface[7ce95aec19cf85e]::interface::run_compiler<core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>, rustc_driver_impl[a06908b56a08a5f3]::run_compiler::{closure#0}>::{closure#1}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>::{closure#0}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>
  42:     0x7f5b10b770ea - <<std[f6dbfb092428beee]::thread::Builder>::spawn_unchecked_<rustc_interface[7ce95aec19cf85e]::util::run_in_thread_with_globals<rustc_interface[7ce95aec19cf85e]::util::run_in_thread_pool_with_globals<rustc_interface[7ce95aec19cf85e]::interface::run_compiler<core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>, rustc_driver_impl[a06908b56a08a5f3]::run_compiler::{closure#0}>::{closure#1}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>::{closure#0}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[3cc7221ca1340ad0]::result::Result<(), rustc_span[413b5bd4996aee21]::ErrorGuaranteed>>::{closure#2} as core[3cc7221ca1340ad0]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  43:     0x7f5b0bacac8b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8f96bdbdef28286c
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/alloc/src/boxed.rs:2022:9
  44:     0x7f5b0bacac8b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hdc6d9d278290e1ae
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/alloc/src/boxed.rs:2022:9
  45:     0x7f5b0bacac8b - std::sys::pal::unix::thread::Thread::new::thread_start::hd8b6cbeea7b365c6
                               at /rustc/aed2187d53b8789e3a37f50ae36f894a2a679077/library/std/src/sys/pal/unix/thread.rs:108:17
  46:     0x7f5b0b9d8609 - start_thread
  47:     0x7f5b0b8fb353 - clone
  48:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/playground/rustc-ice-2024-04-28T18_30_30-119.txt` to your bug report

note: compiler flags: --crate-type bin -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [mir_built] building MIR for `foo::{closure#0}::{closure#0}`
#1 [check_unsafety] unsafety-checking `foo`
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `playground` (bin "playground")
```

</p>
</details>

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.
-->
@rustbot modify labels: +regression-from-nightly-to-nightly -regression-untriaged

Thanks for investigating!

### Workaround

`async move || { ... }` -> `move || async move { ... }`

--- a/compiler/rustc_hir_typeck/src/upvar.rs
+++ b/compiler/rustc_hir_typeck/src/upvar.rs
@@ -204,6 +204,60 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             fake_reads: Default::default(),
         };
 
+        let _ = euv::ExprUseVisitor::new(
+            &FnCtxt::new(self, self.tcx.param_env(closure_def_id), closure_def_id),
+            &mut delegate,
+        )
+        .consume_body(body);
+
+        // There are several curious situations with coroutine-closures where
+        // analysis is too aggressive with borrows when the coroutine-closure is
+        // marked `move`. Specifically:
+        //
+        // 1. If the coroutine-closure was inferred to be `FnOnce` during signature
+        // inference, then it's still possible that we try to borrow upvars from
+        // the coroutine-closure because they are not used by the coroutine body
+        // in a way that forces a move. See the test:
+        // `async-await/async-closures/force-move-due-to-inferred-kind.rs`.
+        //
+        // 2. If the coroutine-closure is forced to be `FnOnce` due to the way it
+        // uses its upvars, but not *all* upvars would force the closure to `FnOnce`.
+        // See the test: `async-await/async-closures/force-move-due-to-actually-fnonce.rs`.
+        //
+        // This would lead to an impossible to satisfy situation, since `AsyncFnOnce`
+        // coroutine bodies can't borrow from their parent closure. To fix this,
+        // we force the inner coroutine to also be `move`. This only matters for
+        // coroutine-closures that are `move` since otherwise they themselves will
+        // be borrowing from the outer environment, so there's no self-borrows occuring.
+        //
+        // One *important* note is that we do a call to `process_collected_capture_information`
+        // to eagerly test whether the coroutine would end up `FnOnce`, but we do this
+        // *before* capturing all the closure args by-value below, since that would always
+        // cause the analysis to return `FnOnce`.
+        if let UpvarArgs::Coroutine(..) = args
+            && let hir::CoroutineKind::Desugared(_, hir::CoroutineSource::Closure) =
+                self.tcx.coroutine_kind(closure_def_id).expect("coroutine should have kind")
+            && let parent_hir_id =
+                self.tcx.local_def_id_to_hir_id(self.tcx.local_parent(closure_def_id))
+            && let parent_ty = self.node_ty(parent_hir_id)
+            && let hir::CaptureBy::Value { move_kw } =
+                self.tcx.hir_node(parent_hir_id).expect_closure().capture_clause
+        {
+            // (1.) Closure signature inference forced this closure to `FnOnce`.
+            if let Some(ty::ClosureKind::FnOnce) = self.closure_kind(parent_ty) {
+                capture_clause = hir::CaptureBy::Value { move_kw };
+            }
+            // (2.) The way that the closure uses its upvars means it's `FnOnce`.
+            else if let (_, ty::ClosureKind::FnOnce, _) = self
+                .process_collected_capture_information(
+                    capture_clause,
+                    &delegate.capture_information,
+                )
+            {
+                capture_clause = hir::CaptureBy::Value { move_kw };
+            }
+        }
+
         // As noted in `lower_coroutine_body_with_moved_arguments`, we default the capture mode
         // to `ByRef` for the `async {}` block internal to async fns/closure. This means
         // that we would *not* be moving all of the parameters into the async block by default.
@@ -253,34 +307,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        let _ = euv::ExprUseVisitor::new(
-            &FnCtxt::new(self, self.tcx.param_env(closure_def_id), closure_def_id),
-            &mut delegate,
-        )
-        .consume_body(body);
-
-        // If a coroutine is comes from a coroutine-closure that is `move`, but
-        // the coroutine-closure was inferred to be `FnOnce` during signature
-        // inference, then it's still possible that we try to borrow upvars from
-        // the coroutine-closure because they are not used by the coroutine body
-        // in a way that forces a move.
-        //
-        // This would lead to an impossible to satisfy situation, since `AsyncFnOnce`
-        // coroutine bodies can't borrow from their parent closure. To fix this,
-        // we force the inner coroutine to also be `move`. This only matters for
-        // coroutine-closures that are `move` since otherwise they themselves will
-        // be borrowing from the outer environment, so there's no self-borrows occuring.
-        if let UpvarArgs::Coroutine(..) = args
-            && let hir::CoroutineKind::Desugared(_, hir::CoroutineSource::Closure) =
-                self.tcx.coroutine_kind(closure_def_id).expect("coroutine should have kind")
-            && let parent_hir_id =
-                self.tcx.local_def_id_to_hir_id(self.tcx.local_parent(closure_def_id))
-            && let parent_ty = self.node_ty(parent_hir_id)
-            && let Some(ty::ClosureKind::FnOnce) = self.closure_kind(parent_ty)
-        {
-            capture_clause = self.tcx.hir_node(parent_hir_id).expect_closure().capture_clause;
-        }
-
         debug!(
             "For closure={:?}, capture_information={:#?}",
             closure_def_id, delegate.capture_information
@@ -289,7 +315,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         self.log_capture_analysis_first_pass(closure_def_id, &delegate.capture_information, span);
 
         let (capture_information, closure_kind, origin) = self
-            .process_collected_capture_information(capture_clause, delegate.capture_information);
+            .process_collected_capture_information(capture_clause, &delegate.capture_information);
 
         self.compute_min_captures(closure_def_id, capture_information, span);
 
@@ -545,13 +571,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     fn process_collected_capture_information(
         &self,
         capture_clause: hir::CaptureBy,
-        capture_information: InferredCaptureInformation<'tcx>,
+        capture_information: &InferredCaptureInformation<'tcx>,
     ) -> (InferredCaptureInformation<'tcx>, ty::ClosureKind, Option<(Span, Place<'tcx>)>) {
         let mut closure_kind = ty::ClosureKind::LATTICE_BOTTOM;
         let mut origin: Option<(Span, Place<'tcx>)> = None;
 
         let processed = capture_information
-            .into_iter()
+            .iter()
+            .cloned()
             .map(|(place, mut capture_info)| {
                 // Apply rules for safety before inferring closure kind
                 let (place, capture_kind) =

--- a/tests/ui/async-await/async-closures/force-move-due-to-actually-fnonce.rs
+++ b/tests/ui/async-await/async-closures/force-move-due-to-actually-fnonce.rs
@@ -1,0 +1,27 @@
+//@ aux-build:block-on.rs
+//@ edition:2021
+//@ check-pass
+
+#![feature(async_closure)]
+
+extern crate block_on;
+
+fn consume(_: String) {}
+
+fn main() {
+    block_on::block_on(async {
+        let x = 1i32;
+        let s = String::new();
+        // `consume(s)` pulls the closure's kind down to `FnOnce`,
+        // which means that we don't treat the borrow of `x` as a
+        // self-borrow (with `'env` lifetime). This leads to a lifetime
+        // error which is solved by forcing the inner coroutine to
+        // be `move` as well, so that it moves `x`.
+        let c = async move || {
+            println!("{x}");
+            // This makes the closure FnOnce...
+            consume(s);
+        };
+        c().await;
+    });
+}

--- a/tests/ui/async-await/async-closures/force-move-due-to-inferred-kind.rs
+++ b/tests/ui/async-await/async-closures/force-move-due-to-inferred-kind.rs
@@ -1,0 +1,24 @@
+//@ aux-build:block-on.rs
+//@ edition:2021
+//@ check-pass
+
+#![feature(async_closure)]
+
+extern crate block_on;
+
+fn force_fnonce<T: async FnOnce()>(t: T) -> T { t }
+
+fn main() {
+    block_on::block_on(async {
+        let x = 1i32;
+        // `force_fnonce` pulls the closure's kind down to `FnOnce`,
+        // which means that we don't treat the borrow of `x` as a
+        // self-borrow (with `'env` lifetime). This leads to a lifetime
+        // error which is solved by forcing the inner coroutine to
+        // be `move` as well, so that it moves `x`.
+        let c = force_fnonce(async move || {
+            println!("{x}");
+        });
+        c().await;
+    });
+}

